[GeoMechanicsApplication] Add output item for the shear capacity

avdg81 opened the pull request:
**📝 Description**

When using a Mohr-Coulomb material model, the shear capacity can be output on the user's request by adding `"GEO_SHEAR_CAPACITY"` to the requested integration point results. This works for diff order U-Pw elements as well as non-diff order ones.

Other changes include:

- Introduced constitutive law utility functions for retrieving the cohesion and friction angle. These work for UDSM material models as well as the new Mohr-Coulomb model with a tension cut-off.
- Changed the file extension of a header file: `.hpp` => `.h`

avdg81 committed:
Changed the file extension of a header file

avdg81 committed:
Added a generic getter for the cohesion

avdg81 committed:
Added a generic getter for the friction angle

avdg81 committed:
Eliminated some duplicated code

avdg81 committed:
Reordered some logic

avdg81 committed:
U-Pw small strain elements can now calculate shear capacity values

Other changes include:
- Define, create, and register a new application variable named `GEO_SHEAR_CAPACITY`.
- Added a stub constitutive law for testing purposes.
- Added a unit test for the case where the Mohr-Coulomb model parameters are supplied using the `UMAT_PARAMETERS`, `INDEX_OF_UMAT_C_PARAMETER`, and `INDEX_OF_UMAT_PHI_PARAMETER` properties.

avdg81 committed:
The k0 procedure now retrieves "phi" using the stress-strain utility function

avdg81 committed:
Added two unit tests for fetching the cohesion

One test covers the case where the cohesion is supplied through the `UMAT_PARAMETERS` item; the other one uses the `GEO_COHESION` item.

avdg81 committed:
Added two unit tests for fetching the friction angle

One test covers the case where the friction angle is supplied through the `UMAT_PARAMETERS` item; the other one uses the `GEO_FRICTION_ANGLE` item.

avdg81 committed:
Modified a unit test

Use modified input values to not exactly arrive at shear capacity values of 1.0.

avdg81 committed:
Simplified a unit test

Use `GEO_COHESION` and `GEO_FRICTION_ANGLE` rather than `UMAT_PARAMETERS` and two indices.

avdg81 committed:
The U-Pw diff order element can now calculate the shear capacity at its integration points

Also added a test helper that creates a 6-noded triangle.

avdg81 committed:
Added the first integration test

This test consists of a single 4-noded U-Pw element with a Mohr-Coulomb material model. All degrees of freedom have been prescribed. The test checks both the (prescribed) displacement field as well as the shear capacity values.

avdg81 committed:
Added the second integration test

This test consists of a single 8-noded U-Pw diff order element with a Mohr-Coulomb material model. All degrees of freedom have been prescribed. The test checks both the (prescribed) displacement field as well as the shear capacity values.

avdg81 committed:
Added new integration tests to the "small" test suite

avdg81 committed:
Added a class for output utilities

So far, this class implements calculation of the shear capacity values.

avdg81 committed:
Updated the first integration test

Prescribe the displacement vector at each node individually, just as for the second integration test.

avdg81 committed:
Create test elements using a utility function

Added a utility function that creates a 2D 6-noded U-Pw diff order element.

avdg81 committed:
Removed an unused static member function

avdg81 committed:
Several minor changes

- Added a blank line.
- Used ternary operators to simplify some code.
- Removed two unused `#include`s.
- Added one more unit test to improve the code coverage.

avdg81 committed:
One of the new tests only works for debug builds

avdg81 committed:
Added documentation of the integration tests

avdg81 committed:
Minor corrections to the test documentation

avdg81 committed:
Added a figure showing the test setup

Also made a minor modification to the test documentation.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -11,7 +11,30 @@
+double GetValueOfUMatParameter(const Properties& rProperties, const Variable<int>& rIndexVariable)
+{
+    KRATOS_ERROR_IF_NOT(rProperties.Has(UMAT_PARAMETERS) && rProperties.Has(rIndexVariable))
+        << "Material " << rProperties.Id() << " does not have UMAT_PARAMETERS and/or "
```
The && condition indicates that both UMAT_PARAMETERS and rIndexVariable should be present. We could be specific about which one(s) is\are missing.

avdg81 replied:
Good point. Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/output_utilities.hpp:
```diff
@@ -0,0 +1,42 @@
+                                             const Properties&            rProperties)
+    {
+        const auto c   = ConstitutiveLawUtilities::GetCohesion(rProperties);
+        const auto phi = ConstitutiveLawUtilities::GetFrictionAngle(rProperties);
```
The capacity can be computed for any constitutive law as long as we also provided a C and Phi. These C and Phi are not necessarily used in the constitutive law and therefore not necessarily check on bounds. Still I would like them to obey 0 < c and  0 <= phi < 90. Can we think of a natural place to check this once.

avdg81 replied:
I would say that $`c`$ and $`\phi`$ should be checked by either the constitutive law of which they are part or by the material input reader. I wouldn't want to make it the responsibility of `GetCohesion` and `GetFrictionAngle`. At most, we could build in debug checks to see whether the values are within bounds. But in all honesty, I would be in favor of a solution where the convention is that all `Properties` instances contain validated data only, to avoid the necessity to check the data over and over again. Perhaps we can discuss this in person?

**Update:** we have discussed this point. Another, potentially better way of validating the input needed for calculating an output item is to override the `Check` member function of the output process. II will add a placeholder issue on the backlog to discuss this further.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_small_strain_u_pw_diff_order_element.cpp:
```diff
@@ -0,0 +1,52 @@
+    p_element->CalculateOnIntegrationPoints(GEO_SHEAR_CAPACITY, actual_shear_capacity_values, dummy_process_info);
+
+    // Assert
+    KRATOS_EXPECT_VECTOR_NEAR(actual_shear_capacity_values, (Vector{ScalarVector{3, 0.75}}),
```
Is the () around vector{ScalarVector{3, 0.75}} necessary?

avdg81 replied:
Yes, unfortunately it is. My understanding is that the macro parser may sometimes get confused when using brace-initialization (i.e. `{}`). By enclosing the expression in a pair of parentheses, we overcome that problem.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_u_pw_small_strain_element.cpp:
```diff
@@ -537,4 +539,31 @@ KRATOS_TEST_CASE_IN_SUITE(UPwSmallStrainElement_SetValuesOnIntegrationPointsMatr
+    p_element->CalculateOnIntegrationPoints(GEO_SHEAR_CAPACITY, actual_shear_capacity_values, dummy_process_info);
+
+    // Assert
+    KRATOS_EXPECT_VECTOR_NEAR(actual_shear_capacity_values, (Vector{ScalarVector{3, 0.75}}),
```
Same question about necessity of the () pair around vector{ScalarVector(

avdg81 replied:
See my previous answer.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/single_element_with_Mohr_Coulomb/README.md:
```diff
@@ -0,0 +1,22 @@
+
+In both tests, all degrees of freedom have been prescribed.  In vertical direction, the elements are compressed by imposing a vertical displacement of the element's top edge by $`-0.015 \mathrm{m}`$.  In horizontal direction, the elements are extended by displacing the right edge by $`+0.015 \mathrm{m}`$.  The water pressure degrees of freedom have all been set to $`0.0 \mathrm{Pa}`$.
+
+The elastic behavior of the elements is described by a Young's modulus of $`100 \mathrm{Pa}`$ and a Poisson's ratio of $`0.0`$.  The material behavior is described using a Mohr-Coulomb model with a cohesion $`c`$ of $`2.0 \mathrm{Pa}`$ and a friction angle $`\phi = 0.0^{\circ}`$.  The material model has been implemented by the shared library `MohrCoulomb64.dll`.
```
This implies that shear capacity should never become larger that 1.

avdg81 replied:
Yes, I think that is correct. Or do I overlook something?

WPK4FEM commented on applications/GeoMechanicsApplication/tests/single_element_with_Mohr_Coulomb/SmallStrainUPwDiffOrderElement2D8N/mesh.mdpa:
```diff
@@ -0,0 +1,133 @@
+    1
+  End SubModelPartElements
+  Begin SubModelPartConditions
+  End SubModelPartConditions
```
Maybe remove all these empty entries for tables, Conditions, Elements etc. to keep the files shorter.

avdg81 replied:
Good point. Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/single_element_with_Mohr_Coulomb/UPwSmallStrainElement2D4N/ProjectParameters.json:
```diff
@@ -0,0 +1,149 @@
+        "newmark_beta":                       0.25,
+        "newmark_gamma":                      0.5,
+        "newmark_theta":                      0.5,
+        "strategy_type":                      "linear",
```
This blocks the nonlinear behaviour of the constitutive law if it would try to go beyond plastic limits.

avdg81 replied:
Good catch. I have changed the `"strategy_type"` to `"newton_raphson"` (in both tests). Let me know if you'd prefer some other strategy.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/single_element_with_Mohr_Coulomb/UPwSmallStrainElement2D4N/ProjectParameters.json:
```diff
@@ -0,0 +1,149 @@
+        "block_builder":                      true,
+        "solution_type":                      "Quasi-Static",
+        "scheme_type":                        "Newmark",
+        "reset_displacements":                false,
```
The lines above and below this reset_displacements line all give information for the Newmark scheme. Preferably the reset_displacements command  is placed before or after this block.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/single_element_with_Mohr_Coulomb/UPwSmallStrainElement2D4N/ProjectParameters.json:
```diff
@@ -0,0 +1,149 @@
+        "compute_reactions":                  true,
+        "move_mesh_flag":                     false,
+        "reform_dofs_at_each_step":           false,
+        "nodal_smoothing":                    false,
```
I think these are the defaults and do not need repetition here.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/single_element_with_Mohr_Coulomb/UPwSmallStrainElement2D4N/mesh.mdpa:
```diff
@@ -0,0 +1,77 @@
+
+Begin SubModelPart all
+  Begin SubModelPartTables
+  End SubModelPartTables
```
See empty blocks remark for previous mdpa file

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_single_element_with_Mohr_Coulomb.py:
```diff
@@ -0,0 +1,96 @@
+    This class contains a test for a single element with a Mohr-Coulomb material model.
+    """
+
+    def assert_displacement_field(
```
Why check a displacement field that is completely prescribed. This is an input == output check, where the purpose of the test lies more in the shear capacity check. The test looks to be designed using a desired principal stress field and adapting the displacement field to those. Hence, I find it more logical to check if the displacement field leads to those principal stresses and if then with these principal stresses the shear capacity is computed correctly.

avdg81 replied:
I had noticed that applying the correct Dirichlet boundary conditions wasn't as straightforward as I had expected. Therefore, I have included this check. But in the end, what matters for the shear capacity calculation is to have a stress field that matches our expectations. So your suggestion to also check the computed stress field makes sense. I have added that check.

WPK4FEM reviewed (commented):
Dear Anne,

Thank you for fixing and making the shear capacity check available. Not sure if all my remarks here fall within the scope of this PR> Please have a look.

Regards, Wijtze Pieter

avdg81 committed:
Processed first review suggestion

Split a combined check into two separate checks to get more specific error messages.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -41,4 +66,18 @@ void ConstitutiveLawUtilities::SetConstitutiveParameters(ConstitutiveLaw::Parame
+                                         : GetValueOfUMatParameter(rProperties, INDEX_OF_UMAT_C_PARAMETER);
+}
+
+double ConstitutiveLawUtilities::GetFrictionAngle(const Properties& rProperties)
```
I think it would help to rename this to GetFrictionAngleInDegrees. In other parts of the code we rely on this, so I'd make it specific

avdg81 replied:
Agreed. It helps to be specific about the applicable angular unit. Done.

avdg81 committed:
Removed several empty tables from `mdpa` files

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -11,7 +11,32 @@
+        << "Material " << rProperties.Id() << " does not have " << rIndexVariable.Name() << "\n";
+
+    const auto index = rProperties[rIndexVariable]; // 1-based index
+    KRATOS_DEBUG_ERROR_IF(index < 1 ||
```
What is the reason this is a debug error?

avdg81 replied:
It is not the responsibility of this function to check that the provided index is in range. Such checks should have been carried out by the owning constitutive law. However, if that input check is somehow omitted, it would still be nice to check this precondition when using a debug build. I don't want the users of our software (who use release builds) to pay for repeatedly checking the index. Let me know if you feel that we should do something differently here.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -41,4 +66,18 @@ void ConstitutiveLawUtilities::SetConstitutiveParameters(ConstitutiveLaw::Parame
     rConstitutiveParameters.SetDeterminantF(detF);
 }
+
+double ConstitutiveLawUtilities::GetCohesion(const Properties& rProperties)
+{
+    return rProperties.Has(GEO_COHESION) ? rProperties[GEO_COHESION]
+                                         : GetValueOfUMatParameter(rProperties, INDEX_OF_UMAT_C_PARAMETER);
+}
```
For wrong material parameters (i.e. an accidental GEO_COHESION added that's different from the cohesion in the UMAT_PARAMETERS), we get (quite hard to find) inconsistencies. Is there a check done somewhere for this (e.g. in the constitutive laws)?

avdg81 replied:
Not that I'm aware of. It is a valid point, but I also feel this is out of scope for this PR. I've discussed the issue of improved input checking with @WPK4FEM as well, and I'll add a new issue to our product backlog to account for it. Let me know if you want something else.

avdg81 committed:
Changed the `"strategy_type"` to `"newton_raphson"`

This would allow for correct behavior when the test would be modified to simulate plastic behavior, too.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_processes/test_apply_c_phi_reduction_process.cpp:
Apparently, I forgot to format here, thanks for doing so :)

avdg81 replied:
You're welcome.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/stub_constitutive_law.h:
```diff
@@ -0,0 +1,26 @@
+class StubConstitutiveLaw : public ConstitutiveLaw
+{
+public:
+    bool                                   RequiresInitializeMaterialResponse() override;
```
We could use google mock here, but I leave that up to you

avdg81 replied:
Yes, you are right. I understand that from a promotional point of view it would be desirable to start using GMock more frequently. So I have modified the class such that it now uses GMock. Thanks for the suggestion.

avdg81 committed:
Cleaned up the `ProjectParameters.json` files of the new tests

Removed duplications of default input parameters.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_utilities/element_setup_utilities.h:
```diff
@@ -37,6 +37,10 @@ class ElementSetupUtilities
                                               const Properties::Pointer& rProperties);
     static Element::Pointer Create2D6NElement();
 
+    static Element::Pointer Create2D6NDiffOrderElement(const PointerVector<Node>& rNodes,
+                                                       const Properties::Pointer& rProperties);
+    static Element::Pointer Create2D6NDiffOrderElement();
```
We could also make the utilities templated (with an element type), but I leave it up to you if you want to do that now (it's a bit out of scope)

avdg81 replied:
I like this idea, but I think it goes beyond the scope of this PR. Let's create a follow-up PR for this.

rfaasse commented on applications/GeoMechanicsApplication/tests/single_element_with_Mohr_Coulomb/README.md:
```diff
@@ -0,0 +1,22 @@
+
+The prescribed displacement field results in the following stress field: $`\sigma_{xx} = +1.5 \mathrm{Pa}`$, $`\sigma_{yy} = -1.5 \mathrm{Pa}`$, and $`\sigma_{zz} = 0.0 \mathrm{Pa}`$.  All shear stress components are equal to $`0.0`$.  Note that only elastic deformations occur.
+
+Both tests assert the prescribed displacement fields (to ensure that the correct Dirichlet boundary conditions have been applied).  In addition, these tests assert the shear capacity, which equals $`0.75`$ in each integration point.
```
Could we add a simple explanation for why it should be 0.75, or is that hard to do?

avdg81 replied:
I have attempted to explain how the value has been calculated. Let me know what you think.

rfaasse replied:
Nice, thanks!

rfaasse reviewed (commented):
Nice addition of functionality by adding nicely reusable generic utilities! I have a couple of comments/suggestions, but none of them are major things

avdg81 committed:
Added a check for the stress field

avdg81 committed:
Processed Richard's review suggestions

- Renamed a member function to make clear that the returned angle is in degrees.
- Use GMock for the stub constitutive law.
- Explained why the shear capacity equals 0.75.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/stub_constitutive_law.h:
```diff
@@ -13,14 +13,24 @@
+    StubConstitutiveLaw()
+    {
+        ON_CALL(*this, RequiresInitializeMaterialResponse).WillByDefault(testing::Return(false));
+        ON_CALL(*this, Clone).WillByDefault([]() { return MakeStubConstitutiveLaw(); });
```
I'd propose to simplify this, but I'd only do that if there's something else that also needs to change (not worth waiting on the pipelines once more in my opinion)
```suggestion
        ON_CALL(*this, Clone).WillByDefault([]() { return std::make_shared<StubConstitutiveLaw> });
```

avdg81 replied:
Hmm... that's odd. I've tried exactly this before and then it didn't work. I thought it had to do with attempting to construct an instance of `StubConstitutiveLaw` from its constructor. But now that I've tried it again (to see which error message I got exactly) it all works as expected. As a result, we no longer need the implementation file (`.cpp`), since it had become empty. :-)

rfaasse reviewed (dismissed):
Thanks for processing the review comments! In my opinion this is ready to go! I have one minor suggestion, but I wouldn't implement it unless there is another change needed anyway (not worth the pipeline wait).

avdg81 committed:
Removed a redundant helper function

It turned out that the helper function could be inlined after all. As a result. an implementation file could be removed, since it had become empty.

WPK4FEM reviewed (approved):
Thank you for taking up the review comments.
Good to go for me.

rfaasse reviewed (approved):
Good to go!
